site/content/blog/2020-11-17-mysql-index-merge.md

```diff
@@ -0,0 +1,178 @@
++----+-------------+-------+------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+------+------+----------+----------------------------------+
+1 row in set (0.06 sec)
+```
+
+This was taking several minutes to run in their environment, and apparently, if they `USE INDEX (idx_posts_channel_id_delete_at_create_at)`, it runs in less than a second. From the above `EXPLAIN` output, we can see that it is choosing the `idx_posts_create_at` index, whereas choosing `idx_posts_channel_id_delete_at_create_at` is clearly the better one.
+
+Before we dive any further, let us look at the table schema and layout.
+
+```sql
+mysql> show CREATE TABLE Posts;
+CREATE TABLE `Posts` (
+  `Id` varchar(26) NOT NULL,
+  `CreateAt` bigint(20) DEFAULT NULL,
+  `UpdateAt` bigint(20) DEFAULT NULL,
+  `EditAt` bigint(20) DEFAULT NULL,
+  `DeleteAt` bigint(20) DEFAULT NULL,
+  `IsPinned` tinyint(1) DEFAULT NULL,
+  `UserId` varchar(26) DEFAULT NULL,
+  `ChannelId` varchar(26) DEFAULT NULL,
+  `RootId` varchar(26) DEFAULT NULL,
+  `Message` text,
+  PRIMARY KEY (`Id`),
+  KEY `idx_posts_update_at` (`UpdateAt`),
+  KEY `idx_posts_create_at` (`CreateAt`),
+  KEY `idx_posts_delete_at` (`DeleteAt`),
+  KEY `idx_posts_channel_id` (`ChannelId`),
+  KEY `idx_posts_root_id` (`RootId`),
+  KEY `idx_posts_user_id` (`UserId`),
+  KEY `idx_posts_channel_id_update_at` (`ChannelId`,`UpdateAt`),
+  KEY `idx_posts_channel_id_delete_at_create_at` (`ChannelId`,`DeleteAt`,`CreateAt`),
+  FULLTEXT KEY `idx_posts_message_txt` (`Message`),
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
+```
+
+This is an abridged version of the `Posts` table for brevity. But it contains the essential elements for us to understand the problem. Our main columns of interest are `CreateAt`, `DeleteAt`, `UpdateAt`, and `ChannelId`. Each of them have their individual indices, and there are 2 additional multi-column indices. One having `ChannelId` and `UpdateAt`. Another involving `ChannelId`, `DeleteAt`, and `CreateAt`.
+
+Now let's look into the query again.
+
+```sql
+SELECT Id FROM Posts WHERE ChannelId = 'x' AND DeleteAt = y AND CreateAt < z
+	ORDER BY CreateAt DESC
+	LIMIT 1
+```
+
+We are filtering the table with 3 columns- `ChannelId`, `DeleteAt`, and `CreateAt`. And then ordering by `CreateAt` and just getting the first row. It becomes clear why choosing the multi-column index is faster than choosing the one for `CreateAt`. Simply because the query filters with all 3 columns which leads to scanning a much smaller dataset. But then, why is MySQL choosing the wrong index in the first place? Using `USE INDEX` was the nuclear button which I didn't want to use unless there were no other options.
```

Comment:
```suggestion
We're filtering the table with three columns: `ChannelId`, `DeleteAt`, and `CreateAt`. And then ordering by `CreateAt` and just getting the first row. It becomes clear why choosing the multi-column index is faster than choosing the one for `CreateAt`. Simply because the query filters with all three columns which leads to scanning a much smaller dataset. But then, why is MySQL choosing the wrong index in the first place? Using `USE INDEX` was the nuclear button which I didn't want to use unless there were no other options.
```